interbtc-spec/docs/source/spec/oracle.rst

```diff
@@ -143,104 +128,104 @@ Specification
 
 *Postconditions*
 
-* The ``SatoshiPerBytes`` MUST be set to the provided ``btcTxFeesPerByte``.
+For each ``(oracleKey, value)`` pair,
 
-.. _getExchangeRate:
+* ``RawValuesUpdated[oracleKey]`` MUST be set to true
+* ``RawValues[oracleKey]`` MUST be set to a ``TimeStamped`` values, where,
 
-getExchangeRate
+  * ``TimeStamped.timestamp`` MUST be the current time,
+  * ``TimeStamped.value`` MUST be ``value``.
+
+.. _getPrice:
+
+getPrice
 ---------------
 
-Returns the latest exchange rate, as received from the external data sources.
+Returns the latest medianized value for the given key, as calculated from the received external data sources.
 
 Specification
 .............
 
 *Function Signature*
 
-``getExchangeRate()``
+``getPrice(oracleKey)``
+
+*Parameters*
+
+* ``oracleKey``: the key for which the value should be returned
 
 *Preconditions*
 
-* The ``LastExchangeRateTime`` MUST NOT be before the current time minus the ``MaxDelay``.
+* ``EchangeRate[oracleKey]`` MUST NOT be ``None``. That is, sufficient oracles must have submitted unexpired values.
 
 *Postconditions*
 
-* MUST return the fixed point base exchange rate.
+* MUST return the fixed point value for the given key.
+
 
-.. _getLastExchangeRateTime:
+.. _oracle_onInitialize:
 
-getLastExchangeRateTime
-------------------------
+onInitialize
+---------------
 
-Returns the UNIX timestamp of when the last exchange rate was received from the external data sources.
+This function is called at the start of every block. When new values have been submitted, or when old values expire, this function update the aggregate value.
 
 Specification
 .............
 
 *Function Signature*
 
-``getLastExchangeRateTime()``
+``onInitialize()``
 
 *Postconditions*
 
-* MUST return the 32-bit UNIX timestamp.
+* If ``RawValuesUpdated`` is empty, i.e., ``feedValues`` was not yet called since the initialization of the parachain, then the ``OracleOffline`` MUST be set in the :ref:`security` pallet.
+* For each ``(oracleKey, updated)`` in ``RawValuesUpdated``, if ``updated`` is true, or the current time is greater than ``ValidUntil[oracle]``,
+
+  * ``RawValuesUpdated[oracleKey]`` MUST be set to false
+  * ``ExchangeRate[oracleKey]`` MUST be set to the middle value of the sorted list of unexpired values from ``RawValues[oracleKey]``. If there are an even number, one MAY be arbitrarily picked.
+  * ``ValidUntil[oracleKey]`` MUST be set to ``MaxDelay`` plus the minimum timestamp from the unexpired values in ``RawValues[oracleKey]``.
 
+.. TODO: recover_from_oracle_offline
```

Comment:
I think we can reference the security module here right?

Reply:
Oracle offline detection is still a work in progress